src/audio/host_zephyr.c

```diff
@@ -0,0 +1,1135 @@
+		return NULL;
+	dev->ipc_config = *config;
+
+	hd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*hd));
+	if (!hd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, hd);
+	hd->ipc_host = *ipc_host;
+
+	/* request HDA DMA with shared access privilege */
+	dir = hd->ipc_host.direction == SOF_IPC_STREAM_PLAYBACK ?
+		DMA_DIR_HMEM_TO_LMEM : DMA_DIR_LMEM_TO_HMEM;
+
+	hd->dma = dma_get(dir, 0, DMA_DEV_HOST, DMA_ACCESS_SHARED);
+	if (!hd->dma) {
+		comp_err(dev, "host_new(): dma_get() returned NULL");
+		rfree(hd);
+		rfree(dev);
+		return NULL;
+	}
+
+	/* init buffer elems */
+	dma_sg_init(&hd->config.elem_array);
+	dma_sg_init(&hd->host.elem_array);
+	dma_sg_init(&hd->local.elem_array);
+
+	ipc_build_stream_posn(&hd->posn, SOF_IPC_STREAM_POSITION, dev->ipc_config.id);
+
+	hd->msg = ipc_msg_init(hd->posn.rhdr.hdr.cmd, sizeof(hd->posn));
+	if (!hd->msg) {
+		comp_err(dev, "host_new(): ipc_msg_init failed");
+		dma_put(hd->dma);
+		rfree(hd);
+		rfree(dev);
+		return NULL;
+	}
+
+	hd->chan = NULL;
+	hd->copy_type = COMP_COPY_NORMAL;
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static void host_free(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	comp_info(dev, "host_free()");
+
+	dma_put(hd->dma);
+
+	ipc_msg_free(hd->msg);
+	dma_sg_free(&hd->config.elem_array);
+	rfree(hd);
+	rfree(dev);
+}
+
+static int host_elements_reset(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_elem *source_elem = NULL;
+	struct dma_sg_elem *sink_elem = NULL;
+	struct dma_sg_elem *local_elem;
+
+	/* setup elem to point to first source elem */
+	source_elem = hd->source->elem_array.elems;
+	if (source_elem) {
+		hd->source->current = 0;
+		hd->source->current_end = source_elem->src + source_elem->size;
+	}
+
+	/* setup elem to point to first sink elem */
+	sink_elem = hd->sink->elem_array.elems;
+	if (sink_elem) {
+		hd->sink->current = 0;
+		hd->sink->current_end = sink_elem->dest + sink_elem->size;
+	}
+
+	/* local element */
+	if (source_elem && sink_elem) {
+		local_elem = hd->config.elem_array.elems;
+		local_elem->dest = sink_elem->dest;
+		local_elem->size =
+			dev->direction == SOF_IPC_STREAM_PLAYBACK ?
+			sink_elem->size : source_elem->size;
+		local_elem->src = source_elem->src;
+	}
+
+	return 0;
+}
+
+static int host_verify_params(struct comp_dev *dev,
+			      struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_dbg(dev, "host_verify_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "host_verify_params(): comp_verify_params() failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+/* configure the DMA params and descriptors for host buffer IO */
+static int host_params(struct comp_dev *dev,
+		       struct sof_ipc_stream_params *params)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_config *config = &hd->config;
+	struct dma_sg_elem *sg_elem;
+	struct dma_config *dma_cfg = &hd->z_config;
```

Comment:
The main issue is dma_config, Im still using old config and translating it to the dma_config at the end, which is obviously wrong but the old config contain some fields which are not available in the new config, like is_scheduling_source, not sure if we can ignore such fields?